src/crimson/common/logclient.cc

```diff
@@ -0,0 +1,363 @@
+  string graylog_port_str = get_str_map_key(log_to_graylog_port, log_channel,
+					    &CLOG_CONFIG_DEFAULT_KEY);
+  int graylog_port = atoi(graylog_port_str.c_str());
+
+  set_log_to_monitors(to_monitors);
+  set_log_to_syslog(to_syslog);
+  set_syslog_facility(syslog_facility);
+  set_log_prio(prio);
+
+  if (to_graylog && !graylog) { /* should but isn't */
+    graylog = seastar::make_shared<ceph::logging::Graylog>("clog");
+  } else if (!to_graylog && graylog) { /* shouldn't but is */
+    graylog = nullptr;
+  }
+
+  if (to_graylog && graylog) {
+    graylog->set_fsid(fsid);
+    graylog->set_hostname(host);
+  }
+
+  if (graylog && (!graylog_host.empty()) && (graylog_port != 0)) {
+    graylog->set_destination(graylog_host, graylog_port);
+  }
+
+  logger().debug("{} to_monitors: {} to_syslog: {}"
+	  "syslog_facility: {} prio: {} to_graylog: {} graylog_host: {}"
+	  "graylog_port: {}", __func__, (to_monitors ? "true" : "false"),
+	  (to_syslog ? "true" : "false"), syslog_facility, prio,
+	  (to_graylog ? "true" : "false"), graylog_host, graylog_port);
+}
+
+void LogChannel::do_log(clog_type prio, std::stringstream& ss)
+{
+  while (!ss.eof()) {
+    string s;
+    getline(ss, s);
+    if (!s.empty()) {
+      do_log(prio, s);
+    }
+  }
+}
+
+void LogChannel::do_log(clog_type prio, const std::string& s)
+{
+  if (CLOG_ERROR == prio) {
+    logger().error("log {} : {}", prio, s);
+  } else {
+    logger().warn("log {} : {}", prio, s);
+  }
+  LogEntry e;
+  e.stamp = ceph_clock_now();
+  e.addrs = parent->get_myaddrs();
+  e.name = parent->get_myname();
+  e.rank = parent->get_myrank();
+  e.prio = prio;
+  e.msg = s;
+  e.channel = get_log_channel();
+
+  // seq and who should be set for syslog/graylog/log_to_mon
+  // log to monitor?
+  if (log_to_monitors) {
+    e.seq = parent->queue(e);
+  } else {
+    e.seq = parent->get_next_seq();
+  }
+
+  // log to syslog?
+  if (do_log_to_syslog()) {
+    logger().warn("{} log to syslog", __func__);
+    e.log_to_syslog(get_log_prio(), get_syslog_facility());
+  }
+
+  // log to graylog?
+  if (do_log_to_graylog()) {
+    logger().warn("{} log to graylog", __func__);
+    graylog->log_log_entry(&e);
+  }
+}
+
+MessageURef LogClient::get_mon_log_message(bool flush)
+{
+  if (flush) {
+    if (log_queue.empty()) {
+      return nullptr;
+    }
+    // reset session
+    last_log_sent = log_queue.front().seq;
+  }
+  return _get_mon_log_message();
+}
+
+bool LogClient::are_pending() const
+{
+  return last_log > last_log_sent;
+}
+
+MessageURef LogClient::_get_mon_log_message()
+{
+  if (log_queue.empty()) {
+    return {};
+  }
+
+  // only send entries that haven't been sent yet during this mon
+  // session!  monclient needs to call reset_session() on mon session
+  // reset for this to work right.
+
+  if (last_log_sent == last_log) {
+    return {};
+  }
+
+  // limit entries per message
+  int64_t num_unsent = last_log - last_log_sent;
+  int64_t num_send;
```

Comment:
Why signed integers?

Reply:
I would strongly vote for following the "core guidelines" (and many other sources): only use
unsigned for bit manipulation. All else should be signed.

Reply:
Ah, it was already discussed a few comments ago. I'm reposting your link to the guidelines: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-signed.

This rule has surprised me so much taking into account the overwhelming proliferation of the `std::size_t` across the standard library. [The Google's StyleGuide has thrown some light](https://google.github.io/styleguide/cppguide.html#Integer_Types):

> Because of historical accident, the C++ standard also uses unsigned integers to represent the size of containers - many members of the standards body believe this to be a mistake, but it is effectively impossible to fix at this point. 

Well, this just replaces the question with another one. if `std::size_t` there is just a mistake, what should be used instead? Any potential solution I can imagine would be (even more?) evil:

* a signed variant of `std::size_t` with same `sizeof` value. This would be inconsistent with the underlying hardware's memory model. On x86 `std::size_t` reflects the limitation driven by the segmentation. Somebody can have an array as big as this restriction imposes. Spending one bit for the sign would half the maximum length and bring a nasty undefined behavior like in the case of `std::ptrdiff_t`.
* `ssize_t` (only from `-1` to some positive maximum) could be implemented with a bias. This would deal with the case of `container.size() - 1` but what about `- 2`, `- 3`, etc.?
* signed integer with `sizeof` value bigger than `sizeof(std::size_t)`. This is the worst IMHO as it would violate the zero-overhead principle.

Do you know how this dilemma has been tackled? I'm asking just from curiosity and for the sake of self-improvement. I think we should really stick to the guidelines.

Reply:
Not sure I see the problem (and see, for example, the standard body discussion in https://cplusplus.github.io/LWG/issue2251).
size_t, if I got this right, is only required to be large enough for the largest object allowed by the execution environment. This is alway much smaller than the full AS range. Thus, nothing is really lost when switching to ssize_t (we do not need that last bit)

Reply:
> size_t, if I got this right, is only required to be large enough for the largest object allowed by the execution environment. This is alway much smaller than the full AS range.

I think this isn't universally true :-(. On 64 bits platforms there is little of reason to worry about but 32 bit ones were (and still are) a different story: https://godbolt.org/z/38fM1eWe1.

> Thus, nothing is really lost when switching to ssize_t (we do not need that last bit)

> Not sure I see the problem (...).

It's about an inconsistency (a potential one; I'm still not sure my understanding is correct) between guidelines and higher-level, fundamental language design principles. C++ got me used to the intuition that when there is a clash between efficiency and safety, efficiency wins. I'm not arguing in favor of the former (to be honest, I think it tends to be overrated for most parts of most projects); I'm just worried by its surprising nature.